### PR TITLE
Battle Group Spawning + AI

### PR DESCRIPTION
**Pull request by some-evil-kitty**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no  |
| Needs wipe? | no  |

### Content:
- [x] Server Event
- [x] LAMBS function
- [x] Battlegroup tagging

### Successfully tested on:
- [x] Local MP
- [ ] Dedicated MP

### CODE REVIEW

File: Missionframework/functions/fn_LAMBS_enableReinforcements.sqf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+params [
+    ["_grp", grpNull, [grpNull]]
+];
+
+if (isNull _grp) exitwith {
+    ["Attempted to set LAMBS Reinforcements on null group.", "ERROR"] remoteExecCall ["KPLIB_fnc_log", 2];
+};
+
+(leader _grp) setVariable ["lambs_danger_dangerRadio", true,true];
```
Comment (veteran29) [inline Missionframework/functions/fn_LAMBS_enableReinforcements.sqf:26]:
```suggestion
(leader _grp) setVariable ["lambs_danger_dangerRadio", true, true];
```

File: Missionframework/functions/fn_spawnVehicle.sqf [hunk 2]
```diff
@@ -83,14 +83,16 @@ if (_classname in KPLIB_o_militiaVehicles) then {
...
-        _x addMPEventHandler ["MPKilled", {
-            params ["_unit", "_killer"];
-            ["KPLIB_manageKills", [_unit, _killer]] call CBA_fnc_localEvent;
-        }];
+    _x addEventHandler ["Killed", {
+	params ["_unit", "_killer", "_instigator", "_useEffects"];
+
+     ["KPLIB_manageKills", [_unit,_killer]] call CBA_fnc_serverEvent;
+    }];
+
```
Comment (veteran29) [inline Missionframework/functions/fn_spawnVehicle.sqf:92]: This undoes the changes we've agreed on in Kill Manager PR.
```suggestion
        _x addMPEventHandler ["MPKilled", {
            params ["_unit", "_killer"];
            ["KPLIB_manageKills", [_unit, _killer]] call CBA_fnc_localEvent;
        }];
```

File: Missionframework/functions/fn_spawnVehicle.sqf [hunk 3]
```diff
@@ -83,14 +83,16 @@ if (_classname in KPLIB_o_militiaVehicles) then {
...
-        }];
+    _x addEventHandler ["Killed", {
+	params ["_unit", "_killer", "_instigator", "_useEffects"];
+
+     ["KPLIB_manageKills", [_unit,_killer]] call CBA_fnc_serverEvent;
+    }];
+    
     } forEach _crew;
 };
+};
```
Comment (veteran29) [inline Missionframework/functions/fn_spawnVehicle.sqf:95]: Rogue brace.
```suggestion
};
```

File: Missionframework/scripts/server/ai/battlegroup_ai.sqf [hunk 4]
```diff
@@ -3,51 +3,32 @@ params [
...
-sleep (5 + (random 5));
-reset_battlegroups_ai = false;
-
-if (!((units _grp) isEqualTo []) && (KPLIB_endgame == 0)) then {
-    [_grp] spawn battlegroup_ai;
-};
+{ deleteWaypoint _x } forEachReversed waypoints _grp;
+{_x doFollow leader _grp} forEach units _grp;
+
+_startpos = getPos (leader _grp);
```
Comment (veteran29) [inline Missionframework/scripts/server/ai/battlegroup_ai.sqf:18]: `_startpos` is fetched few lines above
```suggestion
```

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf [hunk 5]
```diff
@@ -28,18 +28,18 @@ if !(_spawn_marker isEqualTo "") then {
...
 
+        for "_i" from 1 to _squadNumber do {
         // Create infantry groups with up to 8 units per squad
-        private _grp = createGroup [KPLIB_side_enemy, true];
-        for "_i" from 0 to (_target_size - 1) do {
-            if (_i > 0 && {(_i % 8) isEqualTo 0}) then {
-                _bg_groups pushBack _grp;
-                _grp = createGroup [KPLIB_side_enemy, true];
-            };
+        private _grp = createGroup [kplib_side_enemy, true];
```
Comment (veteran29) [inline Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:35]:
```suggestion
        private _grp = createGroup [KPLIB_side_enemy, true];
```

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf [hunk 6]
```diff
@@ -28,18 +28,18 @@ if !(_spawn_marker isEqualTo "") then {
...
+        for "_i" from 0 to 7 do {
             [selectRandom _infClasses, markerPos _spawn_marker, _grp] call KPLIB_fnc_createManagedUnit;
         };
-        [_grp] spawn battlegroup_ai;
-        _bg_groups pushBack _grp;
+        [_grp] call KPLIB_fnc_LAMBS_enableReinforcements;
+        [_grp] call battlegroup_ai;
+        _grp setVariable ["KPLIB_isBattleGroup",true];
+        };
```
Comment (veteran29) [inline Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:42]:
```suggestion
            [_grp] call KPLIB_fnc_LAMBS_enableReinforcements;
            [_grp] call battlegroup_ai;
            _grp setVariable ["KPLIB_isBattleGroup",true];
        };
```

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf [hunk 7]
```diff
@@ -55,8 +55,8 @@ if !(_spawn_marker isEqualTo "") then {
             sleep 0.5;
 
             (crew _vehicle) joinSilent _nextgrp;
-            [_nextgrp] spawn battlegroup_ai;
-            _bg_groups pushback _nextgrp;
+            [_nextgrp] call battlegroup_ai;
+            _nextgrp setVariable ["KPLIB_isBattleGroup",true];
```
Comment (veteran29) [inline Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:59]:
```suggestion
            _nextgrp setVariable ["KPLIB_isBattleGroup", true];
```

**Review by veteran29 — CHANGES_REQUESTED** [review]

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf [hunk 8]
```diff
@@ -28,18 +28,18 @@ if !(_spawn_marker isEqualTo "") then {
...
         // Create infantry groups with up to 8 units per squad
         private _grp = createGroup [KPLIB_side_enemy, true];
-        for "_i" from 0 to (_target_size - 1) do {
-            if (_i > 0 && {(_i % 8) isEqualTo 0}) then {
-                _bg_groups pushBack _grp;
-                _grp = createGroup [KPLIB_side_enemy, true];
-            };
+        for "_i" from 0 to 7 do {
             [selectRandom _infClasses, markerPos _spawn_marker, _grp] call KPLIB_fnc_createManagedUnit;
         };
```
Comment (veteran29) [inline Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:38]:
```suggestion
            for "_i" from 0 to 7 do {
                [selectRandom _infClasses, markerPos _spawn_marker, _grp] call KPLIB_fnc_createManagedUnit;
            };
```

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf [hunk 9]
```diff
@@ -28,18 +28,18 @@ if !(_spawn_marker isEqualTo "") then {
 
         // Adjust target size for infantry
         _target_size = 12 max (_target_size * 4);
+        private _squadNumber = round (_target_size/8);
 
+        for "_i" from 1 to _squadNumber do {
         // Create infantry groups with up to 8 units per squad
         private _grp = createGroup [KPLIB_side_enemy, true];
```
Comment (veteran29) [inline Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:35]:
```suggestion
            // Create infantry groups with up to 8 units per squad
            private _grp = createGroup [KPLIB_side_enemy, true];
```

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf [hunk 10]
```diff
@@ -55,8 +55,8 @@ if !(_spawn_marker isEqualTo "") then {
             sleep 0.5;
 
             (crew _vehicle) joinSilent _nextgrp;
-            [_nextgrp] spawn battlegroup_ai;
-            _bg_groups pushback _nextgrp;
+            [_nextgrp] call battlegroup_ai;
+            _nextgrp setVariable ["KPLIB_isBattleGroup", true];
```
Comment (veteran29) [inline Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:60]: You've removed the following line, this means the battlegroup won't be transferred to HC at end of this script. I don't think this is intentional.
```suggestion
_bg_groups pushback _nextgrp;

```

**Review by veteran29 — CHANGES_REQUESTED** [review]
One behaviour question.

Please try to keep indentation consistent in the codebase as it makes it easier to do code reviews.

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf (hunk 10 above)
Comment (some-evil-kitty) [reply Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:60]: That is indeed not intentional. Good catch!

File: Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf [hunk 11]
```diff
@@ -55,8 +55,9 @@ if !(_spawn_marker isEqualTo "") then {
...
+_bg_groups pushback _nextgrp;
```
Comment (veteran29) [inline Missionframework/scripts/server/battlegroup/spawn_battlegroup.sqf:60]: Mu suggestion did not have indentation in it, oof.
```suggestion
            _bg_groups pushback _nextgrp;
```

**Review by veteran29 — APPROVED** [review]
LGTM

**Pull request merged**